pkg/executor/distribute.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2025 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"context"
+	"strings"
+
+	"github.com/pingcap/tidb/pkg/domain/infosync"
+	"github.com/pingcap/tidb/pkg/executor/internal/exec"
+	"github.com/pingcap/tidb/pkg/infoschema"
+	"github.com/pingcap/tidb/pkg/meta/model"
+	"github.com/pingcap/tidb/pkg/parser/ast"
+	"github.com/pingcap/tidb/pkg/table/tables"
+	"github.com/pingcap/tidb/pkg/tablecodec"
+	"github.com/pingcap/tidb/pkg/util/chunk"
+	"github.com/pingcap/tidb/pkg/util/codec"
+	"github.com/tikv/pd/client/errs"
+	pdhttp "github.com/tikv/pd/client/http"
+)
+
+var schedulerName = "balance-range-scheduler"
+
+// DistributeTableExec represents a distribute table  executor.
+type DistributeTableExec struct {
+	exec.BaseExecutor
+
+	tableInfo      *model.TableInfo
+	is             infoschema.InfoSchema
+	partitionNames []ast.CIStr
+	rule           ast.CIStr
+	engine         ast.CIStr
+
+	done      bool
+	keyRanges []*pdhttp.KeyRange
+}
+
+// Open implements the Executor Open interface.
+func (e *DistributeTableExec) Open(context.Context) error {
+	ranges, err := e.getKeyRanges()
+	if err != nil {
+		return err
+	}
+	e.keyRanges = ranges
+	return nil
+}
+
+// Next implements the Executor Next interface.
+func (e *DistributeTableExec) Next(ctx context.Context, chk *chunk.Chunk) error {
+	chk.Reset()
+	if e.done {
+		return nil
+	}
+	e.done = true
+	err := e.distributeTable(ctx)
+	if err != nil {
+		return err
+	}
+	config, err := infosync.GetSchedulerConfig(ctx, schedulerName)
+	configs, ok := config.([]any)
+	if !ok {
+		return errs.ErrClientProtoUnmarshal.FastGenByArgs(config)
+	}
+	jobs := make([]map[string]any, 0, len(configs))
+	for _, cfg := range configs {
+		job, ok := cfg.(map[string]any)
+		if !ok {
+			return errs.ErrClientProtoUnmarshal.FastGenByArgs(cfg)
+		}
+		jobs = append(jobs, job)
+	}
+	if err != nil {
+		return err
+	}
+	alias := e.getAlias()
+	jobID := float64(-1)
+	for _, job := range jobs {
+		if job["alias"] == alias && job["engine"] == e.engine.String() && job["rule"] == e.rule.String() && job["status"] != "finish" {
```

Comment:
What if the job finished soon?

Reply:
And if two jobs with same configurations are submitted, is it possible the first one gets the second job's id?